src/Fixer/ControlStructure/SingleExpressionPerLineFixer.php

```diff
@@ -0,0 +1,364 @@
+    }
+
+    protected function createConfigurationDefinition(): FixerConfigurationResolverInterface
+    {
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder('elements', 'Which expression must have one element by line.'))
+                ->setAllowedTypes(['array'])
+                ->setAllowedValues([new AllowedValueSubset([
+                    self::ELEMENTS_ARRAYS,
+                    self::ELEMENTS_ARGUMENTS,
+                    self::ELEMENTS_PARAMETERS,
+                    self::ELEMENTS_CONTROL_STRUCTURES,
+                    self::SWITCH_CASES,
+                    self::MATCH_EXPRESSIONS,
+                ])])
+                ->setDefault([
+                    self::ELEMENTS_ARRAYS,
+                    self::ELEMENTS_ARGUMENTS,
+                    self::ELEMENTS_PARAMETERS,
+                    self::ELEMENTS_CONTROL_STRUCTURES,
+                    self::SWITCH_CASES,
+                    self::MATCH_EXPRESSIONS,
+                ])
+                ->getOption(),
+        ]);
+    }
+
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
+    {
+        $this->fixArrays = \in_array(self::ELEMENTS_ARRAYS, $this->configuration['elements'], true);
+        $this->fixArguments = \in_array(self::ELEMENTS_ARGUMENTS, $this->configuration['elements'], true);
+        $this->fixParameters = \in_array(self::ELEMENTS_PARAMETERS, $this->configuration['elements'], true);
+        $this->fixControlStructures = \in_array(self::ELEMENTS_CONTROL_STRUCTURES, $this->configuration['elements'], true);
+        $this->fixSwitchCases = \in_array(self::SWITCH_CASES, $this->configuration['elements'], true);
+        $this->fixMatch = \in_array(self::MATCH_EXPRESSIONS, $this->configuration['elements'], true);
+
+        $this->processBlock($tokens, 0, $tokens->count() - 1, false);
+    }
+
+    /**
+     * Process tokens from $tokens[$begin] to $tokens[$end] and add missing new lines:
+     * - If the code is not multiline, there is nothing to add
+     * - If the code is multiline, it looks for newline right after the first token,
+     * after every comma, and right before the last token and add them when they are missing.
+     *
+     * Since recursion is used to handle things like array inside array or functions we need
+     * a parameter $shouldAddNewLine to knows if we are currently fixing the structure, which
+     * requires to add newlines, or if we are just looking for the next one to fix.
+     *
+     * @param null|non-empty-array<string> $lineDelimiters
+     * @param null|non-empty-array<int>    $lineDelimitersIndexes
+     */
+    private function processBlock(
+        Tokens $tokens,
+        int $begin,
+        int $end,
+        bool $shouldAddNewLine,
+        ?array $lineDelimiters = null,
+        ?array $lineDelimitersIndexes = null
+    ): int {
+        $tokenAddedCount = 0;
+        if (!$tokens->isPartialCodeMultiline($begin, $end)) {
```

Comment:
Does this mean that a structure that does not directly contain newlines but contains another structure that does will be fixed anyway? e.g.:
```diff
-$a = [foo(
+$a = [
+foo(
     1,
     2,
-)];
+)
+];

Reply:
Yes.

In the example
```
$a = [foo(
    1,
    2,
)];
```
the array is multiline by definition because the open [ and the close ] are not on the same line.

Reply:
Maybe I'm wrong but my interpretation of those "fully multiline" rules has always been that we should only consider newlines that are directly part of the structure we are fixing. So in this example, the `foo()` function must be fully multiline, but not the array itself.

@Wirone @keradus What's your opinion on this?

Reply:
You're right, cf https://www.php-fig.org/psr/psr-12/#47-method-and-function-calls

> A single argument being split across multiple lines (as might be the case with an anonymous function or array) does not constitute splitting the argument list itself.

so I need to fix this...